analyzer/src/main/kotlin/managers/utils/GradleDependencyGraphBuilder.kt

```diff
@@ -0,0 +1,410 @@
+
+    /**
+     * Update the dependency graph by adding the given [dependency], which may be [transitive], for the scope with name
+     * [scopeName]. Use the provided [repositories] to resolve the package if necessary. All the dependencies of this
+     * dependency are processed recursively.
+     */
+    private fun addDependencyToGraph(
+        scopeName: String,
+        dependency: Dependency,
+        repositories: List<RemoteRepository>,
+        transitive: Boolean
+    ): DependencyReference {
+        val identifier = dependency.toIdentifierString()
+        val issues = issuesForDependency(dependency)
+        val index = updateDependencyMappingAndPackages(identifier, dependency, repositories, issues)
+
+        val ref = when (val result = findDependencyInGraph(index, dependency)) {
+            is GraphSearchResultFound -> result.ref
+            is GraphSearchResultNotFound ->
+                insertIntoGraph(
+                    DependencyIndex(index, result.fragmentIndex),
+                    scopeName,
+                    dependency,
+                    repositories,
+                    issues,
+                    transitive
+                )
+            is GraphSearchResultIncompatible ->
+                insertIntoNewFragment(index, scopeName, dependency, repositories, issues, transitive)
+        }
+
+        return updateScopeMapping(scopeName, ref, transitive)
+    }
+
+    /**
+     * Update internal state for the given dependency [id]. Check whether this ID is already known. If not, add it
+     * to the data managed by this instance, resolve the package, and update the [issues] if necessary. Return the
+     * numeric index for this dependency.
+     */
+    private fun updateDependencyMappingAndPackages(
+        id: String,
+        dependency: Dependency,
+        repositories: List<RemoteRepository>,
+        issues: MutableList<OrtIssue>
+    ): Int {
+        val dependencyIndex = dependencyIndexMapping[id]
+        if (dependencyIndex != null) {
+            return dependencyIndex
+        }
+
+        updateResolvedPackages(id, dependency, repositories, issues)
+        val newIndex = dependencyIds.size
+        dependencyIds.add(id)
+        dependencyIndexMapping[id] = newIndex
+        return newIndex
+    }
+
+    /**
+     * Search for the [dependency] with the given [index] in the fragments of the dependency graph. Return a
+     * [GraphSearchResult] that indicates how to proceed with this dependency.
+     */
+    private fun findDependencyInGraph(index: Int, dependency: Dependency): GraphSearchResult =
+        referenceMappings.find { mapping ->
+            mapping[index]?.takeIf { dependencyTreeEquals(it, dependency) } != null
+        }?.let { it[index] }?.let { GraphSearchResultFound(it) } ?: referenceMappings.withIndex()
+            .find { !it.value.containsKey(index) }
+            ?.let { GraphSearchResultNotFound(it.index) } ?: GraphSearchResultIncompatible
```

Comment:
Wow, that's really hard to grasp. Could you maybe make this a regular function and introduce speaking variable names?

Reply:
Simplified, also by extracting another helper function.